explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+(perf-pgo)=
+
+# Profile-Guided Optimization
+
+[Profile-guided optimization](https://en.wikipedia.org/wiki/Profile-guided_optimization) (PGO) or feedback-driven optimization (FDO) are synonyms for the same concept.  The idea is that a binary is compiled with built-in instrumentation which, when executed, will generate a profile.  This profile can be used as input for the compiler in a subsequent build of the same binary, and it will serve as a guide for further optimizations.
+
+It can be hard to do profiling of real world applications. Ideally, the profile should be generated by a representative workload of the program, but it's not always possible to simulate a representative workload.  Moreover, the built-in instrumentation impacts the overall performance of the binary which introduces a performance penalty.
+
+In order to address these problems, nowadays we use tools like `perf` to "observe" what the binary is doing externally (sampling it, by monitoring events using Linux kernel's Performance Monitoring Unit -- PMU), which makes the process more suitable to be used in production environments.  This technique works better than the regular built-in instrumentation, but it still has a few drawbacks that we will expand later.
```

Comment:
"nowadays we use..." might age badly how about "we can use..."

Reply:
Good catch.  Done.